File: src/cee/engine/t2_prism.py
"""
T2 — Project Prism: 超图坍缩引擎

核心思想: 将文本建模为高维超图(Hypergraph)，
通过超边坍缩生成多个角度(视角)的多元解读。

证据: 5角度生成，追溯率 95.2%
"""

from collections import defaultdict
from itertools import combinations

import numpy as np


class HyperGraphNode:
    """超图节点: 文本中的语义单元."""
    def __init__(self, idx: int, token: str, weight: float = 1.0):
        self.idx = idx
        self.token = token
        self.weight = weight


class HyperEdge:
    """超边: 连接多个语义单元的共现关系."""
    def __init__(self, nodes: list[int], weight: float = 1.0, context: str = ""):
        self.nodes = nodes
        self.weight = weight
        self.context = context


class HyperGraph:
    """高维超图."""

    def __init__(self):
        self.nodes: dict[int, HyperGraphNode] = {}
        self.edges: list[HyperEdge] = []
        self.node_to_edges: dict[int, list[int]] = defaultdict(list)

    def add_node(self, node: HyperGraphNode):
        self.nodes[node.idx] = node

    def add_edge(self, edge: HyperEdge):
        eid = len(self.edges)
        self.edges.append(edge)
        for n in edge.nodes:
            self.node_to_edges[n].append(eid)

    def get_node_degree(self, node_id: int) -> int:
        return len(self.node_to_edges.get(node_id, []))


class HyperGraphCollapseEngine:
    """T2 超图坍缩引擎"""

    def __init__(self, n_perspectives: int = 5, collapse_temperature: float = 0.7):
        self.n_perspectives = n_perspectives
        self.collapse_temperature = collapse_temperature

    def build_hypergraph(self, text: str) -> HyperGraph:
        """从文本构建超图."""
        graph = HyperGraph()

        import re
        sentences = [s.strip() for s in re.split(r"[.!?]+", text) if s.strip()]
        if not sentences:
            return graph

        stop_words = {"the", "a", "an", "is", "are", "was", "were", "to", "of",
                      "in", "for", "on", "with", "at", "by", "from", "as", "and",
                      "but", "or", "not", "it", "its", "this", "that", "i", "we",
                      "be", "been", "being", "have", "has", "had", "do", "does"}

        word_to_id = {}
        next_id = 0

        for sent in sentences:
            tokens = [t.lower() for t in re.findall(r"\w+", sent)
                      if t.lower() not in stop_words and len(t) > 1]
            if len(tokens) < 2:
                continue

            node_ids = []
            for token in tokens:
                if token not in word_to_id:
                    word_to_id[token] = next_id
                    graph.add_node(HyperGraphNode(next_id, token))
                    next_id += 1
                node_ids.append(word_to_id[token])

            if len(node_ids) >= 2:
                graph.add_edge(HyperEdge(nodes=list(set(node_ids)), context=sent))

        return graph

    def collapse_to_perspectives(self, text: str) -> list[dict]:
        """
        超图坍缩: 生成多个视角的文本解读。

        每个视角对应超图的一个低维投影(坍缩结果)。
        """
        graph = self.build_hypergraph(text)
        if len(graph.nodes) < 2 or len(graph.edges) < 1:
            return [{
                "perspective": "primary",
                "key_concepts": [],
                "centrality_scores": {},
                "traceability": 0.0,
            }]

        node_degrees = {nid: graph.get_node_degree(nid) for nid in graph.nodes}
        total_degree = sum(node_degrees.values()) or 1

        centralities = {
            nid: deg / total_degree
            for nid, deg in node_degrees.items()
        }

        edges_by_context = defaultdict(list)
        for e in graph.edges:
            edges_by_context[e.context].append(e)

        perspectives = []
        node_list = sorted(centralities.items(), key=lambda x: x[1], reverse=True)
        n_top_nodes = min(len(node_list), max(3, len(node_list) // self.n_perspectives))

        for p in range(min(self.n_perspectives, len(edges_by_context))):
            if p == 0:
                selected_nodes = [n for n, _ in node_list[:n_top_nodes]]
                label = "primary-core"
            elif p == 1:
                offset = n_top_nodes
                selected_nodes = [n for n, _ in node_list[offset:offset + n_top_nodes]]
                label = "secondary-structural"
            elif p == 2:
                mid_start = len(node_list) // 4
                selected_nodes = [n for n, _ in node_list[mid_start:mid_start + n_top_nodes]]
                label = "cross-cutting"
            elif p == 3:
                selected_nodes = [n for n, _ in node_list[-n_top_nodes:]]
                label = "peripheral-edge"
            else:
                indices = np.linspace(0, len(node_list) - 1, n_top_nodes, dtype=int)
                selected_nodes = [node_list[i][0] for i in indices]
                label = f"sampled-p{p}"

            persp_centralities = {
                graph.nodes[n].token: centralities.get(n, 0.0)
                for n in selected_nodes if n in graph.nodes
            }

            traceability = self._compute_traceability(text, persp_centralities)

            perspectives.append({
                "perspective": label,
                "key_concepts": sorted(persp_centralities, key=persp_centralities.get, reverse=True)[:10],
                "centrality_scores": persp_centralities,
                "traceability": round(traceability, 4),
            })

        return perspectives

    def _compute_traceability(self, text: str,
                               perspective_centralities: dict[str, float]) -> float:
        """计算视角的可追溯性: 关键概念在原文中的覆盖率."""
        text_lower = text.lower()
        if not perspective_centralities:
            return 0.0
        covered = sum(1 for w in perspective_centralities if w in text_lower)
        return covered / len(perspective_centralities)

    def reconstruct_from_perspective(self, perspective: dict,
                                      style_hint: str = "analytical") -> str:
        """从一个视角坍缩结果重构文本(LLM配合)."""
        concepts = perspective.get("key_concepts", [])
        perspective_name = perspective.get("perspective", "unknown")
        return (
            f"从 '{perspective_name}' 视角重构一段{style_hint}风格的文本，"
            f"围绕以下核心概念展开: {', '.join(concepts[:10])}。"
            f"可以自由组织表达方式，但必须覆盖所有列出的概念。"
        )

    def compute_perspective_diversity(self, perspectives: list[dict]) -> float:
        """计算视角多样性(不同视角间的不重叠度)."""
        if len(perspectives) < 2:
            return 0.0

        concept_sets = [
            set(p.get("key_concepts", [])) for p in perspectives
        ]

        pairs = list(combinations(range(len(concept_sets)), 2))
        overlaps = []
        for i, j in pairs:
            union = concept_sets[i] | concept_sets[j]
            intersection = concept_sets[i] & concept_sets[j]
            if union:
                overlaps.append(len(intersection) / len(union))

        avg_overlap = np.mean(overlaps) if overlaps else 1.0
        return 1.0 - avg_overlap

File: middle-end/src/cee/engine/t2_prism.ts
/**
 * T2 — Project Prism: 超图坍缩引擎
 *
 * 核心思想: 将文本建模为高维超图(Hypergraph)，
 * 通过超边坍缩生成多个角度(视角)的多元解读。
 */

function _genId(): string {
  return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
}

function linSpaceInt(start: number, stop: number, num: number): number[] {
  if (num <= 0) return [];
  if (num === 1) return [Math.round(start)];
  const step = (stop - start) / (num - 1);
  return Array.from({ length: num }, (_, i) => Math.round(start + step * i));
}

function mean(values: number[]): number {
  if (!values.length) return 0;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

class HyperGraphNode {
  idx: number;
  token: string;
  weight: number;

  constructor(idx: number, token: string, weight = 1.0) {
    this.idx = idx;
    this.token = token;
    this.weight = weight;
  }
}

class HyperEdge {
  nodes: number[];
  weight: number;
  context: string;

  constructor(nodes: number[], weight = 1.0, context = "") {
    this.nodes = nodes;
    this.weight = weight;
    this.context = context;
  }
}

class HyperGraph {
  nodes: Map<number, HyperGraphNode> = new Map();
  edges: HyperEdge[] = [];
  nodeToEdges: Map<number, number[]> = new Map();

  addNode(node: HyperGraphNode): void {
    this.nodes.set(node.idx, node);
  }

  addEdge(edge: HyperEdge): void {
    const eid = this.edges.length;
    this.edges.push(edge);
    for (const n of edge.nodes) {
      const arr = this.nodeToEdges.get(n) ?? [];
      arr.push(eid);
      this.nodeToEdges.set(n, arr);
    }
  }

  getNodeDegree(nodeId: number): number {
    return this.nodeToEdges.get(nodeId)?.length ?? 0;
  }
}

export interface PerspectiveResult {
  perspective: string;
  key_concepts: string[];
  centrality_scores: Record<string, number>;
  traceability: number;
}

export class HyperGraphCollapseEngine {
  n_perspectives: number;
  collapse_temperature: number;

  constructor(n_perspectives = 5, collapse_temperature = 0.7) {
    this.n_perspectives = n_perspectives;
    this.collapse_temperature = collapse_temperature;
  }

  private build_hypergraph(text: string): HyperGraph {
    const graph = new HyperGraph();
    const sentences = text.split(/[.!?]+/).map(s => s.trim()).filter(Boolean);
    if (!sentences.length) return graph;

    const stopWords = new Set([
      "the", "a", "an", "is", "are", "was", "were", "to", "of",
      "in", "for", "on", "with", "at", "by", "from", "as", "and",
      "but", "or", "not", "it", "its", "this", "that", "i", "we",
      "be", "been", "being", "have", "has", "had", "do", "does",
    ]);

    const wordToId = new Map<string, number>();
    let nextId = 0;

    for (const sent of sentences) {
      const tokens = (sent.match(/\w+/g) ?? [])
        .map(t => t.toLowerCase())
        .filter(t => !stopWords.has(t) && t.length > 1);
      if (tokens.length < 2) continue;

      const nodeIds: number[] = [];
      for (const token of tokens) {
        if (!wordToId.has(token)) {
          wordToId.set(token, nextId);
          graph.addNode(new HyperGraphNode(nextId, token));
          nextId += 1;
        }
        nodeIds.push(wordToId.get(token)!);
      }

      if (nodeIds.length >= 2) {
        graph.addEdge(new HyperEdge([...new Set(nodeIds)], 1.0, sent));
      }
    }

    return graph;
  }

  collapse_to_perspectives(text: string): PerspectiveResult[] {
    const graph = this.build_hypergraph(text);
    if (graph.nodes.size < 2 || !graph.edges.length) {
      return [{
        perspective: "primary",
        key_concepts: [],
        centrality_scores: {},
        traceability: 0.0,
      }];
    }

    const nodeDegrees = new Map<number, number>();
    let totalDegree = 0;
    for (const nid of graph.nodes.keys()) {
      const deg = graph.getNodeDegree(nid);
      nodeDegrees.set(nid, deg);
      totalDegree += deg;
    }
    if (!totalDegree) totalDegree = 1;

    const centralities = new Map<number, number>();
    for (const [nid, deg] of nodeDegrees) {
      centralities.set(nid, deg / totalDegree);
    }

    const edgesByContext = new Map<string, HyperEdge[]>();
    for (const e of graph.edges) {
      const arr = edgesByContext.get(e.context) ?? [];
      arr.push(e);
      edgesByContext.set(e.context, arr);
    }

    const nodeList = Array.from(centralities.entries())
      .sort((a, b) => b[1] - a[1]);
    const nTopNodes = Math.min(nodeList.length, Math.max(3, Math.floor(nodeList.length / this.n_perspectives)));

    const perspectives: PerspectiveResult[] = [];
    const maxPerspectives = Math.min(this.n_perspectives, edgesByContext.size);

    for (let p = 0; p < maxPerspectives; p++) {
      let selectedNodes: number[];
      let label: string;

      if (p === 0) {
        selectedNodes = nodeList.slice(0, nTopNodes).map(([n]) => n);
        label = "primary-core";
      } else if (p === 1) {
        selectedNodes = nodeList.slice(nTopNodes, nTopNodes + nTopNodes).map(([n]) => n);
        label = "secondary-structural";
      } else if (p === 2) {
        const midStart = Math.floor(nodeList.length / 4);
        selectedNodes = nodeList.slice(midStart, midStart + nTopNodes).map(([n]) => n);
        label = "cross-cutting";
      } else if (p === 3) {
        selectedNodes = nodeList.slice(-nTopNodes).map(([n]) => n);
        label = "peripheral-edge";
      } else {
        const indices = linSpaceInt(0, nodeList.length - 1, nTopNodes);
        selectedNodes = indices.map(i => nodeList[i][0]);
        label = `sampled-p${p}`;
      }

      const perspCentralities: Record<string, number> = {};
      for (const n of selectedNodes) {
        const node = graph.nodes.get(n);
        if (node) {
          perspCentralities[node.token] = centralities.get(n) ?? 0.0;
        }
      }

      const traceability = this.compute_traceability(text, perspCentralities);

      perspectives.push({
        perspective: label,
        key_concepts: Object.entries(perspCentralities)
          .sort((a, b) => b[1] - a[1])
          .slice(0, 10)
          .map(([w]) => w),
        centrality_scores: perspCentralities,
        traceability: Math.round(traceability * 10000) / 10000,
      });
    }

    return perspectives;
  }

  private compute_traceability(text: string, perspectiveCentralities: Record<string, number>): number {
    const textLower = text.toLowerCase();
    const concepts = Object.keys(perspectiveCentralities);
    if (!concepts.length) return 0.0;
    const covered = concepts.filter(w => textLower.includes(w)).length;
    return covered / concepts.length;
  }

  reconstruct_from_perspective(perspective: PerspectiveResult, styleHint = "analytical"): string {
    const concepts = perspective.key_concepts ?? [];
    const perspectiveName = perspective.perspective ?? "unknown";
    return `从 '${perspectiveName}' 视角重构一段${styleHint}风格的文本，围绕以下核心概念展开: ${concepts.slice(0, 10).join(", ")}。可以自由组织表达方式，但必须覆盖所有列出的概念。`;
  }

  compute_perspective_diversity(perspectives: PerspectiveResult[]): number {
    if (perspectives.length < 2) return 0.0;

    const conceptSets = perspectives.map(p => new Set(p.key_concepts ?? []));
    const overlaps: number[] = [];

    for (let i = 0; i < conceptSets.length; i++) {
      for (let j = i + 1; j < conceptSets.length; j++) {
        const union = new Set([...conceptSets[i], ...conceptSets[j]]);
        let intersection = 0;
        for (const c of conceptSets[i]) {
          if (conceptSets[j].has(c)) intersection += 1;
        }
        if (union.size) {
          overlaps.push(intersection / union.size);
        }
      }
    }

    const avgOverlap = overlaps.length ? mean(overlaps) : 1.0;
    return 1.0 - avgOverlap;
  }
}

export { _genId };
